File: 00_java_c_pointer/java/src/Main.java
public class Main {

    
    
    public static void main(String[] args) {
        int x = 3;
        int x2 = 4;
        Pointer<Integer> y = new Pointer<Integer>(x2);
        
        int result1 = square(y);
        int result2 = square(toPointer(x));
        
        System.out.println(result1);
        System.out.println(result2);
    }
    
    
    static int square(Pointer<Integer> value) {
        return value.dereference() * value.dereference();
    }

    
    
    // JUST A HELPER
    static <T> Pointer<T> toPointer(T value) {
        return new Pointer<>(value);
    }

}

File: 00_java_c_pointer/c/src/main.c
#include <stdio.h>
int square(int* value);
int square(int* value) {
    return (*value) * (*value); // dereferenzierung
}
int main() {
    int x = 3;
    int x2 = 4;
    int* y = &x2; // erstellen Referenz auf die Speicheradresse von x

    int result1 = square(y);
    int result2 = square(&x);

    printf("%i\n", result1);
    printf("%i\n", result2);

}

